### ISSUE TITLE
ipahealthcheck_exporter stopped working after upgrade from Fedora 37 to 39

### ISSUE THREAD

**Issue by dmgeurts** [camptocamp/ipahealthcheck_exporter#15]
The Journal shows the following error:

```
Mar 13 11:35:03 ipa.domain.com systemd[1]: Started ipahealthcheck-exporter.service - Prometheus ipahealthcheck_exporter.
Mar 13 11:35:03 ipa.domain.com ipahealthcheck_exporter[729983]: time="2024-03-13T11:35:03+01:00" level=info msg="Running exporter version development, commit none, built at 2024-03-13 11:35:03.101782656 +0100 CET m=+0.002325668" source="ipahealthcheck_exporter.go:283"
Mar 13 11:35:03 ipa.domain.com ipahealthcheck_exporter[729983]: time="2024-03-13T11:35:03+01:00" level=info msg="ipa-healthcheck exporter listening on http://127.0.0.1:9888\n" source="ipahealthcheck_exporter.go:285"
Mar 13 11:40:00 ipa.domain.com ipahealthcheck_exporter[729983]: time="2024-03-13T11:40:00+01:00" level=info msg="Scraping metrics from /usr/bin/ipa-healthcheck" source="ipahealthcheck_exporter.go:115"
Mar 13 11:40:00 ipa.domain.com ipahealthcheck_exporter[729983]: time="2024-03-13T11:40:00+01:00" level=info msg="using sudo to execute health check" source="ipahealthcheck_exporter.go:126"
Mar 13 11:40:00 ipa.domain.com sudo[733710]: ipahealthcheck-exporter : PWD=/ ; USER=root ; COMMAND=/usr/bin/ipa-healthcheck --source ipahealthcheck.meta.services --output-file /dev/shm/ipa-healthcheck.out1284489533
Mar 13 11:40:02 ipa.domain.com ipahealthcheck_exporter[729983]: time="2024-03-13T11:40:02+01:00" level=fatal msg="Cannot unmarshal json from ipa-healthcheck output: unexpected end of JSON input" source="ipahealthcheck_exporter.go:143"
Mar 13 11:40:02 ipa.domain.com systemd[1]: ipahealthcheck-exporter.service: Main process exited, code=exited, status=1/FAILURE
Mar 13 11:40:02 ipa.domain.com systemd[1]: ipahealthcheck-exporter.service: Failed with result 'exit-code'.
Mar 13 11:40:03 ipa.domain.com systemd[1]: ipahealthcheck-exporter.service: Scheduled restart job, restart counter is at 4.
```
After noticing that ipa-healthcheck now needs to run as root I created a sudo exception for user ipahealthcheck-exporter to run ipa-healthcheck without a password. And added the -sudo flag to ExecStart in the systemd service file.

This resolved the "must be root to run this script" warning/error:

```
Mar 13 10:50:00 ipa.domain.com ipahealthcheck_exporter[695469]: You must be root to run this script.
Mar 13 10:50:00 ipa.domain.com ipahealthcheck_exporter[694623]: time="2024-03-13T10:50:00+01:00" level=info msg="ipa-healthcheck tool returned errors: exit status 1" source="
ipahealthcheck_exporter.go:132"
Mar 13 10:50:00 ipa.domain.com ipahealthcheck_exporter[694623]: time="2024-03-13T10:50:00+01:00" level=fatal msg="Cannot unmarshal json from ipa-healthcheck output: unexpected end of JSON input" source="ipahealthcheck_exporter.go:143"
```

But the "Cannot unmarshal json" error remains. Have I lost a prerequisite package in the OS upgrade or is something else going on?

**Comment by dmgeurts** [comment]
A permission error?

```
user@ipa:~$ sudo -u ipahealthcheck-exporter /usr/local/bin/ipahealthcheck_exporter -address 127.0.0.1 -sudo -v
[sudo] password for user:
INFO[0000] Running exporter version development, commit none, built at 2024-03-13 17:24:40.700698059 +0100 CET m=+0.004399379  source="ipahealthcheck_exporter.go:283"
INFO[0000] ipa-healthcheck exporter listening on http://127.0.0.1:9888  source="ipahealthcheck_exporter.go:285"
INFO[0019] Scraping metrics from /usr/bin/ipa-healthcheck  source="ipahealthcheck_exporter.go:115"
INFO[0019] using sudo to execute health check            source="ipahealthcheck_exporter.go:126"
Output raised PermissionError: [Errno 13] Permission denied: '/dev/shm/ipa-healthcheck.out3726923885'
FATA[0022] Cannot unmarshal json from ipa-healthcheck output: unexpected end of JSON input  source="ipahealthcheck_exporter.go:143"
```

**Comment by dmgeurts** [comment]
And when run without `-sudo`:

```
user@ipa:~$ sudo -u ipahealthcheck-exporter /usr/local/bin/ipahealthcheck_exporter -address 127.0.0.1 -v
INFO[0000] Running exporter version development, commit none, built at 2024-03-13 17:28:46.621750465 +0100 CET m=+0.001540476  source="ipahealthcheck_exporter.go:283"
INFO[0000] ipa-healthcheck exporter listening on http://127.0.0.1:9888  source="ipahealthcheck_exporter.go:285"
INFO[0073] Scraping metrics from /usr/bin/ipa-healthcheck  source="ipahealthcheck_exporter.go:115"

You must be root to run this script.

INFO[0074] ipa-healthcheck tool returned errors: exit status 1  source="ipahealthcheck_exporter.go:132"
FATA[0074] Cannot unmarshal json from ipa-healthcheck output: unexpected end of JSON input  source="ipahealthcheck_exporter.go:143"
```

**Comment by dmgeurts** [comment]
Checking logs I can see the sudo command issued, but though the file in /dev/shm/ is created, it's empty.

```
/var/log/secure:Mar 13 11:50:00 ipa sudo[741240]: ipahealthcheck-exporter : PWD=/ ; USER=root ; COMMAND=/usr/bin/ipa-healthcheck --source ipahealthcheck.meta.services --output-file /dev/shm/ipa-healthcheck.out2609118389
```

**Comment by dmgeurts** [comment]
The temporary file created [here](https://github.com/camptocamp/ipahealthcheck_exporter/blob/d3442865112177a4c1f816c6649778b2d53f566d/ipahealthcheck_exporter.go#L118) ends up with 0600 permissions and owned by the exporter, my hunch is that ipa-healthcheck is unable to write to the tmpfile when run with sudo, and exporter subsequently throws a jq error when the tmpfile is empty.

Would it be worth adding checks for an empty file to avoid the JQ error, and add a check to see if ipa-healthcheck can write to the tmpfile and chmod it to '0666' or '0660 with `chown :root`'?

**Comment by dmgeurts** [comment]
Pull request submitted with a fix for this issue.

**Comment by lconsuegra** [comment]
Hello again, 

I'm sorry but I read your issue+pull request and am not sure if its needed.

I created the ipahealthcheck-exporter with a sudoers rule and tried to run the following command : 
` sudo -u ipahealthcheck-exporter /usr/local/bin/ipahealthcheck_exporter -port 9999 -v -sudo`

And scraping the exporter works fine, and the leftover tempfile has the following permissions :
`-rw-------. ipahealthcheck-exporter ipahealthcheck-exporter ipa-healthcheck.out535945045`

This is normal since the exporter is run by ipahealthcheck-exporter, so the creation of the file is done by this user (and also the write of the checks by 


Basically what the tool does is this : 

-  `sudo -u ipahealthcheck-exporter touch /dev/shm/ipa-healthcheck.outxx`

-  `sudo -u ipahealthcheck-exporter sudo /usr/bin/ipa-healthcheck --source ipahealthcheck.meta.services --output-file /dev/shm/ipa-healthcheck.outxx`

- Read the checks from the output file and return them as metrics.

Do you have any issues with the above commands ? 
Maybe its an issue with you sudoers rule, can you share it ?

**Comment by dmgeurts** [comment]
The sudo rules for ipahealthcheck-exporter: 

```text
user@ipa:~$ sudo cat /etc/sudoers.d/ipahealthcheck-exporter
ipahealthcheck-exporter ALL = NOPASSWD: /usr/bin/ipa-healthcheck
ipahealthcheck-exporter ALL = NOPASSWD: /usr/bin/chown ^root \/dev\/shm\/ipa-healthcheck.out[0-9]+$
ipahealthcheck-exporter ALL = NOPASSWD: /usr/bin/chmod ^\+r \/dev\/shm\/ipa-healthcheck.out[0-9]+$
ipahealthcheck-exporter ALL = NOPASSWD: /usr/bin/rm ^\/dev\/shm\/ipa-healthcheck.out[0-9]+$
```

The following demonstrates the issue (Fedora 39):

```text
user@ipa:~$ sudo -u ipahealthcheck-exporter touch /dev/shm/ipa-healthcheck.outxx
user@ipa:~$ ll /dev/shm
total 3
drwxrwxrwt.  3 root                    root                     100 May 22 19:08 ./
drwxr-xr-x. 20 root                    root                    3780 May 14 10:35 ../
-rw-r--r--.  1 ipahealthcheck-exporter ipahealthcheck-exporter    0 May 22 19:08 ipa-healthcheck.outxx
-rw-------.  1 dirsrv                  dirsrv                    32 May 14 10:35 sem.slapd-DOMAIN-COM.stats
user@ipa:~$ sudo rm /dev/shm/ipa-healthcheck.outxx
user@ipa:~$ ll /dev/shm
total 2
drwxrwxrwt.  3 root    root      80 May 22 19:10 ./
drwxr-xr-x. 20 root    root    3780 May 14 10:35 ../
-rw-------.  1 dirsrv  dirsrv    32 May 14 10:35 sem.slapd-DOMAIN-COM.stats
user@ipa:~$ sudo -u ipahealthcheck-exporter sudo /usr/bin/ipa-healthcheck --source ipahealthcheck.meta.services --output-file /dev/shm/ipa-healthcheck.outxx
user@ipa:~$ ll /dev/shm
total 3
drwxrwxrwt.  3 root    root     100 May 22 19:10 ./
drwxr-xr-x. 20 root    root    3780 May 14 10:35 ../
-rw-r--r--.  1 root    root    3317 May 22 19:10 ipa-healthcheck.outxx
-rw-------.  1 dirsrv  dirsrv    32 May 14 10:35 sem.slapd-DOMAIN-COM.stats
```

**Comment by lconsuegra** [comment]
Alright, I did not had the same behavior so i search a bit and found that its due to a change in the kernel after 4.19 for added security (permissions are restricted in world writable sticky directories, cf [protected-regular](https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-regular)).

The right way to manage that now seems to create a working directory for the user/process in /dev/shm. That way we don't have to harcode sudo in the code and you won't have multiple lines in sudoers (the only one needed is to launch ipa-healthcheck tool).

I have prepared a new version of the exporter, can you test it to see if that works on your environment please ?
[ipahealthcheck_exporter- v0.0.17-rc1.zip](https://github.com/camptocamp/ipahealthcheck_exporter/files/15417842/ipahealthcheck_exporter-.v0.0.17-rc1.zip)

Regards.

**Comment by dmgeurts** [comment]
Sorry for the delay.

That seems to work great so far, thank you!
